### ISSUE TITLE
Using connection in multiple async methods

### ISSUE THREAD

**Issue by HamidTheGeek** [mysql-net/MySqlConnector#758]
Hi there,

I am facing with invalid operation exception when I use the same connection in multiple async methods. The error message is `Cannot set MySqlCommand.CommandText when there is an open DataReader for this command; it must be closed first.`.

The simplified version off my code is as follows.

```csharp
class Name
{
    public async Task MainMethod()
    {
        using var connection = new MySqlConnection(ConnectionString);
        await connection.OpenAsync();
        using var transaction = connection.BeginTransaction(ReadUncommitted)

        if(await IsValid(connection, transaction))
        {
            // Some logic
            await Save(connection, transaction);
        }

        transaction.Commit();
        connection.Close()
    }

    private async Task<bool> IsValue(MySqlConnection connection)
    {
        var command = connection.CreateCommand();
        command.CommandText = "Select Count(*) From Table Where Condition";

        var count = (int) await command.ExecuteScalarAsync();

        return count == 0;
    }

    private async Task Save(MySqlConnection connection)
    {
        var command = connection.CreateCommand();
        // Error happens in the following line
        command.CommandText = "Update Table ....";

        await command.ExecuteNonQueryAsync();
    }
}
```

Due to this [issue](https://github.com/mysql-net/MySqlConnector/issues/554) I learned that I may not use a connection in multiple threads or use it in parallel tasks. But my code is not multi-threading although I use async-await.

**Comment by bgrainger** [comment]
The source code you provided doesn't compile:

* `IsValid` is misspelled
* A `transaction` is passed to `IsValid` and `Save` but they're not declared as taking two parameters
* `(int) await command.ExecuteScalarAsync()` will give an `InvalidCastException` at runtime
* Various semicolons are missing

If I fix all those errors, I can run this sample code without problems; I don't get the exception you report.

If you can provide a small, self-contained reproduction of the problem, then I can investigate further.

My guess is that you have a `MySqlDataReader` that hasn't been closed/disposed when you try to reuse the connection; this is not allowed. Make sure all calls to `ExecuteReader` are in a `using` statement.

**Comment by HamidTheGeek** [comment]
Thank you, @bgrainger , for your feedback.
Unfortunately, It would be hard for you to trace it as I am using [this framework](https://github.com/Geeksltd/Olive). But if you don't mind looking into all the complexity (You would need to look into the ORM provided by the framework), I can provide the actual code. On the other hand, I found a workaround when I was trying to trace the issue. If I wrap the connection with something like the following sample, the issue disappears.

```csharp
using MySql.Data.MySqlClient;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Text;

namespace Olive.Entities.Data
{
    public class MySqlConnectionWrapper : DbConnection
    {
        readonly MySqlConnection RealConnection = new MySqlConnection();

        public override string ConnectionString
        {
            get => RealConnection.ConnectionString;
            set => RealConnection.ConnectionString = value;
        }

        public override string Database => RealConnection.Database;

        public override string DataSource => RealConnection.DataSource;

        public override string ServerVersion => RealConnection.ServerVersion;

        public override ConnectionState State => RealConnection.State;

        public override void ChangeDatabase(string databaseName) =>
                RealConnection.ChangeDatabase(databaseName);

        public override void Close() => RealConnection.Close();

        public override void Open() => RealConnection.Open();

        protected override DbTransaction BeginDbTransaction(IsolationLevel isolationLevel) =>
                RealConnection.BeginTransaction(isolationLevel);

        protected override DbCommand CreateDbCommand() =>
                new MySqlCommand { Connection = RealConnection };
    }
}
```

**Comment by bgrainger** [comment]
Can you provide the actual code that calls that framework?

**Comment by HamidTheGeek** [comment]
Please find the project [here](https://github.com/HamidTheGeek/mysqlconnectionissue758) and consider the following steps:

- Find the database dump file int the `DB` folder next to the solution file.
- Change the connection string in the `appsettings.json`.
- Look and [MySql branch in the Olive](https://github.com/Geeksltd/Olive/tree/mysql) framework for the compatible codes.
- Build and replace the Olive files in the `Assemblies` folder next to the solution file.
- With this API `http://localhost:9015/test` you can easily replicate the issue but it may randomly happen.

Once again thank you and let me know if there were any problems to replicate the issue.

**Comment by bgrainger** [comment]
I couldn't reproduce your bug, but I did get an `ObjectDisposedException` from `MySqlCommand`. I haven't dug into this much yet, but it appears that `MySqlDataReader.Dispose` will dispose the `MySqlCommand` it's associated with if and only if `CommandBehavior.CloseConnection` is specified.

I couldn't find a good reason why this is happening. It may have been done for completeness, but feels unnecessary because it's guaranteed that the `MySqlDataReader` can outlive the `MySqlCommand`, so it's safe for a client to dispose the command then return the reader. See https://github.com/mysql-net/AdoNetApiTest/blob/a939576f2ad692c673352a2c9f9462f90ce136d1/src/AdoNet.Specification.Tests/DataReaderTestBase.cs#L22-L36.

It's also not clear to me whether this is having an impact on your problem.

**Comment by bgrainger** [comment]
The `ObjectDisposedException` that I observed should be addressed by 6248de02b0e6e66e7d1f4910c04a168837432372 and c18facba71d3426ee4dce5b0c06c1dd8f3503df6, which shipped in 0.62.0.

Closing this issue as there isn't a minimal repro.
